### PR TITLE
Connection pools for S3

### PR DESCRIPTION
**Pull request by excitoon**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Use a connection pool for S3 connections, controlled by the `s3_max_connections` settings.

### CODE REVIEW

**Comment by excitoon** [comment]
This PR looks okay but needs tuning of tests or configuration because now `maxConnections` is 25 instead of infinity.

File: src/IO/HTTPCommon.cpp [hunk 1]
```diff
@@ -105,23 +105,53 @@ namespace
...
+            , resolve_host(resolve_host_)
         {
         }
     };
 
     class HTTPSessionPool : private boost::noncopyable
     {
     private:
-        using Key = std::tuple<std::string, UInt16, bool>;
+        using Key = std::tuple<std::string, UInt16, bool, std::string, UInt16, bool>;
```
Comment (akuzm) [inline src/IO/HTTPCommon.cpp:154]: Which std::string is which? A struct would be more clear.

**Comment by akuzm** [comment]
So how exactly does this PR achieve a speedup? I've looked at the code but didn't get an idea what it does.

**Comment by akuzm** [comment]
The Yandex synchronization failure is because the internal repo uses some old Poco version which lacks some methods:

```
$(SOURCE_ROOT)/clickhouse/src/IO/HTTPCommon.cpp:122:26: error: no member named 'setProxyProtocol' in 'Poco::Net::HTTPClientSession'
                session->setProxyProtocol(proxy_scheme);
                ~~~~~~~  ^
$(SOURCE_ROOT)/clickhouse/src/IO/HTTPCommon.cpp:123:26: error: no member named 'setProxyTunnel' in 'Poco::Net::HTTPClientSession'
                session->setProxyTunnel(false);
                ~~~~~~~  ^
2 errors generated.
```

**Comment by alexey-milovidov** [comment]
Simply don't use these methods `#if defined(ARCADIA_BUILD)`

File: src/IO/HTTPCommon.cpp (hunk 1 above)
Comment (excitoon) [reply src/IO/HTTPCommon.cpp]: +

**Comment by akuzm** [comment]
Fuzzer https://github.com/ClickHouse/ClickHouse/issues/15889

**Review by akuzm — APPROVED** [review]

**Pull request merged**